opencog/guile/SchemePrimitive.h

```diff
@@ -30,640 +78,231 @@ class PrimitiveEnviron
-
-					(that->*method.v_ti)(t, i);
-					break;
-				}
-				case V_TIDI:
-				{
-					Type t = SchemeSmob::verify_atom_type(scm_car(args), scheme_name, 1);
-					int i = SchemeSmob::verify_int(scm_cadr(args), scheme_name, 2);
-					double d = SchemeSmob::verify_real(scm_caddr(args), scheme_name, 3);
-					int i2 = SchemeSmob::verify_int(scm_cadddr(args), scheme_name, 4);
-					(that->*method.v_tidi)(t, i, d, i2);
-					break;
-				}
-				case V_V:
-				{
-					(that->*method.v_v)();
-					break;
-				}
-				default:
-					throw RuntimeException(TRACE_INFO,
-						 "Unsupported signature: %d\n", signature);
-			}
-			return rc;
-		}
-	protected:
-		virtual const char *get_name(void) { return scheme_name; }
-		virtual const char *get_module(void) { return scheme_module; }
-		virtual size_t get_size(void) { return sizeof (*this); }
-	public:
-
-#define DECLARE_CONSTR_0(SIG, LSIG, RET_TYPE) \
-	SchemePrimitive(const char* module, const char* name, \
-		 RET_TYPE (T::*cb)(void), T *data) \
-	{ \
-		that = data; \
-		method.LSIG = cb; \
-		scheme_module = module; \
-		scheme_name = name; \
-		signature = SIG; \
-		do_register(module, name, 0); /* cb has 0 args */ \
+	typedef R (C::*Method)(Args...);
+	using PlainTuple = std::tuple<typename std::remove_reference<Args>::type...>;
+
+	Method method;
+	C* that;
+	const char *scheme_module;
+	const char *scheme_name;
+
+public:
+	SchemePrimitiveBase(const char* module, const char* name,
+	                    R (C::*cb)(Args...), C *data)
+	{
+		that = data;
+		method = cb;
+		scheme_module = module;
+		scheme_name = name;
+		do_register(module, name, sizeof...(Args));
 	}
 
-#define DECLARE_CONSTR_1(SIG, LSIG, RET_TYPE, ARG_TYPE) \
-	SchemePrimitive(const char* module, const char* name, \
-		 RET_TYPE (T::*cb)(ARG_TYPE), T *data) \
-	{ \
-		that = data; \
-		method.LSIG = cb; \
-		scheme_module = module; \
-		scheme_name = name; \
-		signature = SIG; \
-		do_register(module, name, 1); /* cb has 1 arg */ \
+protected:
+	virtual const char *get_name(void) { return scheme_name; }
+	virtual const char *get_module(void) { return scheme_module; }
+	virtual size_t get_size(void) { return sizeof (*this); }
+
+	// Convert the ith scm argument to its C++ object. The reason the
+	// return type is also present as third argument is to tell gcc
+	// what overloaded version of scm_to to use. It is a bit of a hack
+	// but not that ugly given that the alternative is to use
+	// specialized templates.
+	bool scm_to(SCM args, size_t idx, bool) const
+	{
+		SCM arg = scm_list_ref(args, scm_from_size_t(idx));
+		return scm_to_bool(arg);
+	}
+	size_t scm_to(SCM args, size_t idx, size_t) const
+	{
+		SCM arg = scm_list_ref(args, scm_from_size_t(idx));
+		return SchemeSmob::verify_size(arg, scheme_name, idx);
+	}
+	int scm_to(SCM args, size_t idx, int) const
+	{
+		SCM arg = scm_list_ref(args, scm_from_size_t(idx));
+		return SchemeSmob::verify_int(arg, scheme_name, idx);
+	}
```

Comment:
these three should be removed, no one should be passing bools, ints or size_t types through the API, that's just fundamentally wrong.